service/src/main/java/com/commercetools/pspadapter/payone/SessionHandler.java

```diff
@@ -0,0 +1,202 @@
+        int retryCounter = 0;
+        try {
+            for (; retryCounter < RETRIES_LIMIT; retryCounter++) {
+                try {
+                    return startSession(paymentId);
+                } catch (final ConcurrentModificationException concurrentModificationException) {
+                    if (retryCounter == RETRIES_LIMIT - 1) {
+                        throw concurrentModificationException;
+                    } else {
+                        Thread.sleep(calculateVariableDelay(retryCounter));
+                    }
+                }
+            }
+        } catch (final ConcurrentModificationException concurrentModificationException) {
+            return handleConcurrentModificationException(paymentId, concurrentModificationException);
+        } catch (final NotFoundException | NoCartLikeFoundException e) {
+            return handleNotFoundException(paymentId, retryCounter, e);
+        } catch (final ErrorResponseException e) {
+            return errorResponseHandler(paymentId, retryCounter, e);
+        } catch (final Exception e) {
+            return handleException(paymentId, retryCounter, e);
+        }
+        return handleException(paymentId, retryCounter,
+                new Exception("Unknown workflow error in PaymentHandler#handlePayment"));
+    }
+
+    private PayoneResult startSession(@Nonnull final String paymentId)
+            throws ConcurrentModificationException {
+
+        final PaymentWithCartLike paymentWithCartLike =
+                commercetoolsQueryExecutor.getPaymentWithCartLike(paymentId);
+
+        final String paymentInterface = paymentWithCartLike
+                .getPayment()
+                .getPaymentMethodInfo()
+                .getPaymentInterface();
+
+        if (!payoneInterfaceName.equals(paymentInterface)) {
+            final String errorMessage = format("Wrong payment interface name: expected '%s', found '%s' for the "
+                    + "commercetools Payment with id '%s'.", payoneInterfaceName, paymentInterface, paymentId);
+            return new PayoneResult(HttpStatusCode.BAD_REQUEST_400, errorMessage);
+        }
+        final PaymentMethodInfo paymentMethodInfo = paymentWithCartLike.getPayment().getPaymentMethodInfo();
+        if (!StringUtils.equals(paymentMethodInfo.getMethod(), INVOICE_KLARNA.getKey())) {
+            final String errorMessage = format("The session can only be started for the payment method '%s'",
+                    INVOICE_KLARNA.getKey());
+            return new PayoneResult(HttpStatusCode.BAD_REQUEST_400, errorMessage);
+        }
+        if (paymentWithCartLike.getPayment().getCustom() != null && StringUtils.isNotBlank(paymentWithCartLike.getPayment().getCustom().getFieldAsString(START_SESSION_RESPONSE))) {
+            final String errorMessage = format("The session was already started for the payment '%s'", paymentId);
+            return new PayoneResult(HttpStatusCode.BAD_REQUEST_400, errorMessage);
```

Comment:
Is this really a bad request? Why not just responding OK?

Reply:
https://github.com/commercetools/commercetools-payone-integration/commit/5a7ee69b6d5a8222a1ad9bda68c492df12f173b5#diff-20236be4a7233c3ddcc1542df444c0f09f864e5cc9f8d177db5e2c2cbc552028R132